applications: matter-ble bridge over thread (#16689)
nrfconnect/sdk-nrf: applications/matter_bridge/doc/matter_bridge_description.rst

Configuring the number of Bluetooth LE bridged devices
------------------------------------------------------

.. tabs::

   .. group-tab:: nRF53 DKs

>       You have to consider several factors to decide how many Bluetooth LE bridged devices to support.

Reviewer: Can we put the text that is repeated for both tabs before the tabs section?

      Every Bluetooth LE bridged device uses a separate Bluetooth LE connection, so you need to consider the number of supported Bluetooth LE connections (selected using the :kconfig:option:`CONFIG_BT_MAX_CONN` Kconfig option) when you configure the number of devices.
      Since the Matter stack uses one Bluetooth LE connection for commissioning, the maximum number of connections you can use for bridged devices is one less than is configured using the :kconfig:option:`CONFIG_BT_MAX_CONN` Kconfig option.

      .. tabs::

>          .. group-tab:: Matter-Bridge over wifi configuration

Reviewer: ```suggestion
         .. group-tab:: Matter-Bridge over Wi-Fi configuration
```

            Increasing the number of Bluetooth LE connections affects the RAM usage on both the application and network cores.
            The current maximum number of Bluetooth LE connections that can be selected using the default configuration is ``10``.
            You can increase the number of Bluetooth LE connections if you decrease the size of the Bluetooth LE TX/RX buffers used by the Bluetooth controller, but this will decrease the communication throughput.
            Build the target using the following command in the project directory to enable a configuration that increases the number of Bluetooth LE connections to ``20`` by decreasing the sizes of Bluetooth LE TX/RX buffers:

            .. parsed-literal::
               :class: highlight

               west build -b nrf5340dk/nrf5340/cpuapp -p -- -Dmatter_bridge_SHIELD=nrf7002ek -DSB_CONFIG_WIFI_PATCHES_EXT_FLASH_STORE=y -DSB_CONFIG_DFU_MULTI_IMAGE_PACKAGE_WIFI_FW_PATCH=y -DSB_CONFIG_WIFI_NRF700X=y -Dmcuboot_CONFIG_UPDATEABLE_IMAGE_NUMBER=3 -DCONFIG_BRIDGED_DEVICE_BT=y -DEXTRA_CONF_FILE="overlay-bt_max_connections_app.conf" -Dipc_radio_EXTRA_CONF_FILE="overlay-bt_max_connections_net.conf"

>          .. group-tab:: Matter-Bridge over thread configuration

Reviewer: ```suggestion
         .. group-tab:: Matter-Bridge over Thread configuration
```

            Increasing the number of Bluetooth LE connections affects the RAM usage on both the application and network cores.
            The current maximum number of Bluetooth LE connections that can be selected using the default configuration is ``10``.
            Due to using thread and BLE protocols in this configuration You can't increase the number of Bluetooth LE connections and additionally you have to decrease the size of the Bluetooth LE TX/RX buffers used by the Bluetooth controller, this will decrease the communication throughput.

>             Build the target using the following command in the project directory to enable a configuration that allow to set the number of Bluetooth LE connections to ``10`` by decreasing the sizes of Bluetooth LE TX/RX buffers:

Reviewer: Isn't it that we say here - the default configuration is 10, but you can configure it to use max number which is 10? :smile:

            .. parsed-literal::
               :class: highlight

               west build -b nrf5340dk/nrf5340/cpuapp -p -- -DCONFIG_BRIDGED_DEVICE_BT=y -Dipc_radio_EXTRA_CONF_FILE="overlay-bt_max_connections_net_thread_nrf5340.conf" -DEXTRA_CONF_FILE="overlay-bt_max_connections_app_thread_nrf5340.conf"

   .. group-tab:: nRF54 DKs

      You have to consider several factors to decide how many Bluetooth LE bridged devices to support.
      Every Bluetooth LE bridged device uses a separate Bluetooth LE connection, so you need to consider the number of supported Bluetooth LE connections (selected using the :kconfig:option:`CONFIG_BT_MAX_CONN` Kconfig option) when you configure the number of devices.
      Since the Matter stack uses one Bluetooth LE connection for commissioning, the maximum number of connections you can use for bridged devices is one less than is configured using the :kconfig:option:`CONFIG_BT_MAX_CONN` Kconfig option.

      .. tabs::

         .. group-tab:: Matter-Bridge over wifi configuration

>             Increasing the number of Bluetooth LE connections affects the RAM usage on both the application and network cores.

Reviewer: I can see a lot of duplications in this text between the tabs. I would suggest contacting @wiba-nordic to discuss how to avoid it, as the only things that changes are build command and the connections number.

Reviewer 2: I'll take a look once the doc build finishes. Hopefully there are parts we can move to be outside of the tabs, and maybe some parts we can have as includes.

Configuring the number of Bluetooth LE bridged devices
------------------------------------------------------

.. tabs::

   You have to consider several factors to decide how many Bluetooth LE bridged devices to support.
   Every Bluetooth LE bridged device uses a separate Bluetooth LE connection, so you need to consider the number of supported Bluetooth LE connections (selected using the :kconfig:option:`CONFIG_BT_MAX_CONN` Kconfig option) when you configure the number of devices.

>    Since the Matter stack uses one Bluetooth LE connection for commissioning, the maximum number of connections you can use for bridged devices is one less than is configured using the :kconfig:option:`CONFIG_BT_MAX_CONN` Kconfig option.

Reviewer 2: This part needs to be outside of the tabs. It disappears from the documentation once you select one of the DK tabs since it's not inside any of them.

>    .. group-tab:: nRF53 DKs

Reviewer 2: For this tab level, something before the inner tabs needs to be added to each tab. Since the only difference is going to be which configurations are supported, I think something like this works:
"The way you configure the number of Bluetooth LE bridged devices depends on the configuration you are using."

The only downside is that it doesn't really work for the 70 series, and I'm not sure what to add there instead.

      .. tabs::

         .. group-tab:: Matter-Bridge over Wi-Fi configuration

            Increasing the number of Bluetooth LE connections affects the RAM usage on both the application and network cores.

>             The current maximum number of Bluetooth LE connections that can be selected using the default configuration is ``10``.

Reviewer 2: These two lines also stay constant for all tabs, so they can be moved completely outside of the tabs.

            You can increase the number of Bluetooth LE connections if you decrease the size of the Bluetooth LE TX/RX buffers used by the Bluetooth controller, but this will decrease the communication throughput.
            Build the target using the following command in the project directory to enable a configuration that increases the number of Bluetooth LE connections to ``20`` by decreasing the sizes of Bluetooth LE TX/RX buffers:

            .. parsed-literal::
               :class: highlight

               west build -b nrf5340dk/nrf5340/cpuapp -p -- -Dmatter_bridge_SHIELD=nrf7002ek -DSB_CONFIG_WIFI_PATCHES_EXT_FLASH_STORE=y -DSB_CONFIG_DFU_MULTI_IMAGE_PACKAGE_WIFI_FW_PATCH=y -DSB_CONFIG_WIFI_NRF700X=y -Dmcuboot_CONFIG_UPDATEABLE_IMAGE_NUMBER=3 -DCONFIG_BRIDGED_DEVICE_BT=y -DEXTRA_CONF_FILE="overlay-bt_max_connections_app.conf" -Dipc_radio_EXTRA_CONF_FILE="overlay-bt_max_connections_net.conf" -DFILE_SUFFIX=nrf70ek

>          .. group-tab:: Matter-Bridge over Thread configuration

Reviewer 3: You mentioned Matter Bridge over Thread only here and in the table below, but it is just a tab, so it is not visible enough. Shouldn't we inform about that somewhere at the bottom of this page, or somewhere else in our docs as well? I have a feeling that this is a bit of hidden information. @kkasperczyk-no what do you think?

Reviewer 3: For example here: 
![image](https://github.com/user-attachments/assets/b55d4a41-2ca7-40ec-b57f-87ebfbd0536e)
there is mention that you can use only Wi-Fi as a transport for Matter in Matter Bridge.

Increasing the number of Bluetooth LE connections affects the RAM usage on both the application and network cores.
The current maximum number of Bluetooth LE connections that can be selected using the default configuration is ``10``.

.. tabs::

> 

Reviewer 2: ```suggestion
You can configure the number of Bluetooth LE bridged devices depending on your device and configuration:
```

             .. parsed-literal::
               :class: highlight

               west build -b nrf54h20dk/nrf54h20/cpuapp -p -- -DSB_CONFIG_Wi-Fi_NRF700X=y -DCONFIG_CHIP_Wi-Fi=y -Dmatter_bridge_SHIELD=nrf700x_nrf54h20dk -DCONFIG_BRIDGED_DEVICE_BT=y -DEXTRA_CONF_FILE="overlay-bt_max_connections_app.conf" -Dipc_radio_EXTRA_CONF_FILE="overlay-bt_max_connections_net.conf"

         .. group-tab:: Matter-Bridge over Thread configuration

            You can increase the number of Bluetooth LE connections if you decrease the size of the Bluetooth LE TX/RX buffers used by the Bluetooth controller, but this will decrease the communication throughput.
            Build the target using the following command in the project directory to enable a configuration that increases the number of Bluetooth LE connections to ``20`` by decreasing the sizes of Bluetooth LE TX/RX buffers:

             .. parsed-literal::
               :class: highlight

               west build -b nrf54h20dk/nrf54h20/cpuapp -p -- -DCONFIG_BRIDGED_DEVICE_BT=y -DEXTRA_CONF_FILE="overlay-bt_max_connections_app.conf" -Dipc_radio_EXTRA_CONF_FILE="overlay-bt_max_connections_net.conf"

>    .. group-tab:: nRF70 DKs

Reviewer 2: ```suggestion
   .. group-tab:: nRF70 DKs
   
      The nRF70 Series supports the Matter bridge over Wi-Fi configuration.
```
Bit of an odd

      .. tabs::

>          .. group-tab:: Matter-Bridge over Wi-Fi configuration

Reviewer 2: ```suggestion
         .. group-tab:: Matter-Bridge over Wi-Fi
```

            You can increase the number of Bluetooth LE connections if you decrease the size of the Bluetooth LE TX/RX buffers used by the Bluetooth controller, but this will decrease the communication throughput.

>             Build the target using the following command in the project directory to enable a configuration that increases the number of Bluetooth LE connections to ``20`` by decreasing the sizes of Bluetooth LE TX/RX buffers:

Reviewer 2: ```suggestion

            Build the target using the following command in the project directory to enable a configuration that increases the number of Bluetooth LE connections to ``20`` by decreasing the sizes of Bluetooth LE TX/RX buffers:
```

   .. group-tab:: nRF53 DKs

      The nRF53 Series supports the Matter bridge over Wi-Fi and Matter bridge over Thread configurations.

      .. tabs::

         .. group-tab:: Matter bridge over Wi-Fi

            You can increase the number of Bluetooth LE connections if you decrease the size of the Bluetooth LE TX/RX buffers used by the Bluetooth controller, but this will decrease the communication throughput.

            Build the target using the following command in the project directory to enable a configuration that increases the number of Bluetooth LE connections to ``20`` by decreasing the size of Bluetooth LE TX/RX buffers:

            .. parsed-literal::
               :class: highlight

               west build -b nrf5340dk/nrf5340/cpuapp -p -- -Dmatter_bridge_SHIELD=nrf7002ek -DSB_CONFIG_WIFI_PATCHES_EXT_FLASH_STORE=y -DSB_CONFIG_DFU_MULTI_IMAGE_PACKAGE_WIFI_FW_PATCH=y -DSB_CONFIG_WIFI_NRF700X=y -Dmcuboot_CONFIG_UPDATEABLE_IMAGE_NUMBER=3 -DCONFIG_BRIDGED_DEVICE_BT=y -DEXTRA_CONF_FILE="overlay-bt_max_connections_app.conf" -Dipc_radio_EXTRA_CONF_FILE="overlay-bt_max_connections_net.conf" -DFILE_SUFFIX=nrf70ek

         .. group-tab:: Matter bridge over Thread

>             You can not increase the default number of Bluetooth LE connections in this configuration.

Reviewer: This paragraph is not quite true. You can't increase the number of supported Bluetooth LE connections using an overlay. But you can always do that by modifying the board files and decreasing the buffer sizes. I would also add a statement here that the default number of supported devices here is 9.

Reviewer 2: Wording out Kamil's suggestion:
```suggestion
            You can not increase the default number of Bluetooth LE connections in this configuration using overlays.
```
And the following after Line 560:
```
            You can still increase the number of connections by modifying the board files and decreasing the buffer sizes.
            The default number of connections is nine.
```

     - Debug version of the application.

       Enables additional features for verifying the application behavior, such as logs.

>    * - Debug
>      - :file:`prj_nrf70ek.conf`
>      - ``nrf70ek``
>      - ``nrf5340dk``
>      - Debug version of the application.
> 
>        Enables Matter-Bridge to work with Wi-Fi.

Reviewer 2: Aligning more with the Thread and Wi-Fi switching entry for the [Matter door lock configurations](https://docs.nordicsemi.com/bundle/ncs-latest/page/nrf/samples/matter/lock/README.html#matter_door_lock_custom_configurations)
![image](https://github.com/user-attachments/assets/d7c252f8-1f79-45a6-a58f-477bf64ba8d8)
```suggestion
   * - Matter bridge over Wi-Fi
     - :file:`prj_nrf70ek.conf`
     - ``nrf70ek``
     - nRF5340 DK with the nRF7002 EK shield attached
     - Debug version of the application with Matter over Wi-Fi enabled.
```
Should also be moved to be the last entry in the list, I think. We don't really have much precedent beyond the Door lock sample, which places the the Debug and Release configurations first.

Is this configuration only used when working with the nRF5340 DK and the shield? If so, that might need to be added to the description.

     - All from `Requirements`_
     - Debug version of the application.

>       Enables additional features for verifying the application behavior, such as logs.

Reviewer 2: ```suggestion
       Enables additional features for verifying the application behavior, such as logs.
```
One missing space manages to break the doc build.

     - Release version of the application.

       Enables only the necessary application functionalities to optimize its performance.

>    * - Matter bridge over Wi-Fi

Reviewer: ```suggestion
   * - Matter bridge over Wi-Fi with nRF5340 DK and nRF7002 EK
```

Reviewer 2: I don't think the devices need to be in the configuration name, in case we add more supported devices later.

Reviewer: But the thing is that there are already more devices that support Matter bridge over Wi-Fi and this particular configuration is related only to nrf5340dk+nrf70ek. Other configurations like nrf54h20+nrf70ek or nrf7002dk are built using other way.

Reviewer 2: Ah, I missed that part. In that case, it makes since to keep the device mentions in both places. 

```suggestion
   * - Matter bridge over Wi-Fi with the nRF5340 DK and nRF7002 EK
```
Note the definitive article.

>      - :file:`prj_nrf70ek.conf`

Reviewer: This probably could be removed.

Reviewer 2: For an empty entry in a table, you can use `---` as the value.

Reviewer: @kg-nordicsemi could you address this one?

Author: It turned out that the prj_nrf70ek.conf file is unnecessary when it comes to building, so I removed it, but I forgot to make changes to the documentation, I'll change it now

     - ``nrf70ek``
     - nRF5340 DK with the nRF7002 EK shield attached
     - Debug version of the application with Matter over Wi-Fi enabled.

>        Enables Matter-Bridge to work with Wi-Fi.

Reviewer: ```suggestion
       Enables the Matter Bridge to work with Wi-Fi on nRF5340 DK.
```

Reviewer 2: ```suggestion
       Enables Matter bridge to work over Wi-Fi.
```
We don't need to specify the device here either, since there's a specific entry for supported devices in the table. We also don't mention the devices for the similar network switching in the Door lock sample.

      .. code-block:: console

>          west build -b nrf54h20dk/nrf54h20/cpuapp -p -- -DSB_CONFIG_Wi-Fi_NRF700X=y -DCONFIG_CHIP_Wi-Fi=y -Dmatter_bridge_SHIELD=nrf700x_nrf54h20dk -DCONFIG_BRIDGED_DEVICE_BT=y -Dmatter_bridge_SNIPPET=onoff_plug

Reviewer: Please verify all configs using WIFI in this readme, as I think that you replaced all WIFI -> Wi-Fi occurrences what broke multiple build commands.
